Move CatBoost evaluation into clickhouse-library-bridge

rschu1ze opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
For security and stability reasons, catboost models are no longer evaluated within the ClickHouse server. Instead, the evaluation was moved into clickhouse-library-bridge, a separate process that loads the catboost library and communicates with the server process via HTTP.

---

This PR moves the catboost model evaluation out of the server process into the library-bridge binary. This serves two goals: On the one hand, crashes / memory corruptions of the catboost library no longer affect the server. On the other hand, we can forbid loading dynamic libraries in the server (catboost was the last consumer of this functionality), thus improving security.

SQL syntax:
```
  SELECT
    FEAT_1, ..., FEAT_N, catboostEvaluate('/path/to/model.bin', FEAT_1, ..., FEAT_N)
  FROM data
  LIMIT 10
```
Required configuration:
```
  <catboost_lib_path>/path/to/libcatboostmodel.so</catboost_lib_path>
```
### Implementation Details

The internal protocol between the server and the library-bridge is simple:

- HTTP GET on path "/extdict_ping":
  A ping, used during the handshake to check if the library-bridge runs.

- HTTP POST on path "extdict_request"
  1. Send a "catboost_GetTreeCount" request from the server to the bridge, containing a library path (e.g /home/user/libcatboost.so) and a model path (e.g. /home/user/model.bin). First, this unloads the catboost library handler associated to the model path (if it was loaded), then loads the catboost library handler associated to the model path, then executes GetTreeCount() on the library handler and finally sends the result back to the server. Step (1) is called once by the server from FunctionCatBoostEvaluate::getReturnTypeImpl(). The library path handler is unloaded in the beginning because it contains state which may no longer be valid if the user runs catboost("/path/to/model.bin", ...) more than once and if "model.bin" was updated in between.
  2. Send "catboost_Evaluate" from the server to the bridge, containing the model path and the features to run the interference on. Step (2) is called multiple times (once per chunk) by the server from function FunctionCatBoostEvaluate::executeImpl(). The library handler for the given model path is expected to be already loaded by Step (1).

Also, the obsolete `modelEvaluate()` function is deleted, as well as statement `SYSTEM RELOAD MODELS` and monitoring view `SYSTEM.SYSTEMMODELS`.

Fixes #27870

alexey-milovidov commented:
More bridge processes complicate building and make the packages larger. Maybe it's reasonable to put it into the library-bridge by extending its API? Note: both variants ok to me.

rschu1ze commented:
Yepp, like discussed in the meeting today, I'll move everything into the library-bridge.

rschu1ze committed:
feat: implement catboost in library-bridge

This commit moves the catboost model evaluation out of the server
process into the library-bridge binary. This serves two goals: On the
one hand, crashes / memory corruptions of the catboost library no longer
affect the server. On the other hand, we can forbid loading dynamic
libraries in the server (catboost was the last consumer of this
functionality), thus improving security.

SQL syntax:

  SELECT
    catboostEvaluate('/path/to/model.bin', FEAT_1, ..., FEAT_N) > 0 AS prediction,
    ACTION AS target
  FROM amazon_train
  LIMIT 10

Required configuration:

  <catboost_lib_path>/path/to/libcatboostmodel.so</catboost_lib_path>

*** Implementation Details ***

The internal protocol between the server and the library-bridge is
simple:

- HTTP GET on path "/extdict_ping":
  A ping, used during the handshake to check if the library-bridge runs.

- HTTP POST on path "extdict_request"
  (1) Send a "catboost_GetTreeCount" request from the server to the
      bridge, containing a library path (e.g /home/user/libcatboost.so) and
      a model path (e.g. /home/user/model.bin). Rirst, this unloads the
      catboost library handler associated to the model path (if it was
      loaded), then loads the catboost library handler associated to the
      model path, then executes GetTreeCount() on the library handler and
      finally sends the result back to the server. Step (1) is called once
      by the server from FunctionCatBoostEvaluate::getReturnTypeImpl(). The
      library path handler is unloaded in the beginning because it contains
      state which may no longer be valid if the user runs
      catboost("/path/to/model.bin", ...) more than once and if "model.bin"
      was updated in between.
  (2) Send "catboost_Evaluate" from the server to the bridge, containing
      the model path and the features to run the interference on. Step (2)
      is called multiple times (once per chunk) by the server from function
      FunctionCatBoostEvaluate::executeImpl(). The library handler for the
      given model path is expected to be already loaded by Step (1).

Fixes #27870

rschu1ze committed:
chore: delete obsolete modelEvaluate() function

- superseded by catboostEvaluate() which no longer uses the internal
  repository for external models

- also removed was statement SYSTEM RELOAD MODELS and the monitoring view
  SYSTEM.SYSTEMMODELS

Commit:
style: fix style

rschu1ze committed:
fix: broken links in documentation (hopefully)

rschu1ze commented on tests/integration/test_catboost_evaluate/test.py:
```diff
@@ -65,41 +74,61 @@ def testModelPathIsNotAConstString(ch_cluster):
+    err = instance.query_and_get_error(
+        "select catboostEvaluate('/etc/clickhouse-server/model/simple_model.bin', 1, 2);"
+    )
+    assert (
```
What an ugly formatting. But I'll better not argue with the autoformatter :smile:

rschu1ze commented on .gitignore:
```diff
@@ -58,6 +58,10 @@ cmake_install.cmake
 CTestTestfile.cmake
 *.a
 *.o
+*.so
```
Noticed that Git does not warn if one wants to check in shared library files.

rschu1ze commented on programs/library-bridge/CatBoostLibraryAPI.h:
```diff
@@ -0,0 +1,49 @@
+#pragma once
```
The code in this file is taken from https://clickhouse.com/codebrowser/ClickHouse/src/Interpreters/CatBoostModel.cpp.html#DB::CatBoostWrapperAPI

rschu1ze commented on programs/library-bridge/CatBoostLibraryHandler.cpp:
```diff
@@ -0,0 +1,376 @@
+    extern const int LOGICAL_ERROR;
+}
+
+CatBoostLibraryHandler::APIHolder::APIHolder(SharedLibrary & lib)
```
Code taken from https://clickhouse.com/codebrowser/ClickHouse/src/Interpreters/CatBoostModel.cpp.html#DB::CatBoostLibHolder

rschu1ze commented on programs/library-bridge/CatBoostLibraryHandler.cpp:
```diff
@@ -0,0 +1,376 @@
+    GetDimensionsCount = lib.tryGet<CatBoostLibraryAPI::GetDimensionsCountFunc>(CatBoostLibraryAPI::GetDimensionsCountName);
+}
+
+CatBoostLibraryHandler::CatBoostLibraryHandler(
```
Code taken from https://clickhouse.com/codebrowser/ClickHouse/src/Interpreters/CatBoostModel.cpp.html#_ZN2DB17CatBoostModelImplC1EPKNS_18CatBoostWrapperAPIERKNSt3__112basic_stringIcNS4_11char_traitsIcEENS4_9allocatorIcEEEE

rschu1ze commented on programs/library-bridge/CatBoostLibraryHandler.cpp:
```diff
@@ -0,0 +1,376 @@
+namespace
+{
+
+/// Buffer should be allocated with features_count * column->size() elements.
```
The code from here until the end of the file is (almost) 1:1 copypasted from https://clickhouse.com/codebrowser/ClickHouse/src/Interpreters/CatBoostModel.cpp.html#_ZNK2DB17CatBoostModelImpl8evaluateERKNSt3__16vectorIPKNS_7IColumnENS1_9allocatorIS5_EEEE

rschu1ze commented on programs/library-bridge/CatBoostLibraryHandlerFactory.cpp:
```diff
@@ -0,0 +1,48 @@
+#include "CatBoostLibraryHandlerFactory.h"
```
The code in this file is similar to https://clickhouse.com/codebrowser/ClickHouse/programs/library-bridge/SharedLibraryHandlerFactory.cpp.html

rschu1ze commented on programs/library-bridge/LibraryBridgeHandlerFactory.cpp:
```diff
@@ -27,12 +27,16 @@ std::unique_ptr<HTTPRequestHandler> LibraryBridgeHandlerFactory::createRequestHa
     {
         if (uri.getPath() == "/extdict_ping")
             return std::make_unique<ExternalDictionaryLibraryBridgeExistsHandler>(keep_alive_timeout, getContext());
+        else if (uri.getPath() == "/catboost_ping")
```
Installs new handler for catboost in the library-bridge.

rschu1ze commented on programs/library-bridge/LibraryBridgeHandlers.cpp:
```diff
@@ -399,4 +410,202 @@ void ExternalDictionaryLibraryBridgeExistsHandler::handleRequest(HTTPServerReque
 }
 
 
+CatBoostLibraryBridgeRequestHandler::CatBoostLibraryBridgeRequestHandler(
```
"Request" and "ping" handlers for catboost in library-bridge. Sort of similar to the external dictionary handlers. More details are mentioned in the commit message.

rschu1ze commented on programs/library-bridge/LibraryBridgeHandlers.cpp:
```diff
@@ -399,4 +410,202 @@ void ExternalDictionaryLibraryBridgeExistsHandler::handleRequest(HTTPServerReque
+
+            const String & model_path = params.get("model_path");
+
+            CatBoostLibraryHandlerFactory::instance().remove(model_path);
```
`remove` followed by `create` may look weird. The reason is explained in the comment in file "programs/library-bridge/LibraryBridgeHandlers.h".

```
///     The library path handler is unloaded in
///     the beginning because it contains state which may no longer be valid if the user runs catboost("/path/to/model.bin", ...) more than
///     once and if "model.bin" was updated in between.
```

rschu1ze commented on src/Functions/catboostEvaluate.cpp:
```diff
@@ -0,0 +1,182 @@
+#include <Interpreters/Context_fwd.h>
+
+
+namespace DB
```
The code in this file is mostly copypasta of https://clickhouse.com/codebrowser/ClickHouse/src/Functions/modelEvaluate.cpp.html

rschu1ze commented on tests/integration/test_catboost_evaluate/config/models_config.xml:
```diff
@@ -0,0 +1,3 @@
+<clickhouse>
+    <catboost_lib_path>/etc/clickhouse-server/model/libcatboostmodel.so</catboost_lib_path>
```
As can be see below, this commit adds three binary files for testing to the Git repository
- "amazon_model.bin" and "simple_model.bin", which are trained catboost models together ca. 5 MB
- "libcatboost.so", 76 MB

"simple_model.bin" and "libcatboost.bin" existed already in the Git repository (they are used by the old modelEvaluate() tests). GitHub only shows them as "added" which is misleading.

I know, checking in binary files into a Git repository is generally not a good habit. But compiling libcatboost.so on-the-fly (which is what the integration tests for external dictionaries in the library-bridge do) is not an option: It uses Yandex's proprietary build system and the build takes too long. 

Oh, and I also got special permission by Nikolai to keep the binary files in Git :smile:

rschu1ze commented:
Tests are finally green.

This PR got rather big ... I have added some comments with context to help with review. Also, a review will hopefully be simpler if one looks at the individual commits individually.

rschu1ze committed:
fix: test 02117_show_create_table_system

kssenii commented on src/BridgeHelper/CatBoostLibraryBridgeHelper.cpp:
```diff
@@ -0,0 +1,121 @@
+        [this, serialized = string_write_buf.str()](std::ostream & os)
+        {
+            os << "model_path=" << escapeForFileName(model_path) << "&";
+            os << "data=" << serialized;
```
hm, as this will go in url and not in post body, it should also be escaped.

kssenii commented on src/Functions/catboostEvaluate.cpp:
```diff
@@ -21,66 +21,80 @@ namespace DB
 
+    void initBridge(const ColumnConst * name_col) const
+    {
+        String library_path = getContext()->getConfigRef().getString("catboost_lib_path");
```
also check `config.has("catboost_lib_path")` and throw meaningful exception? or it is checked already?

kssenii replied:
Yeah, I just thought that we could throw a better exception message that the Poco's, but it is just a suggestion, current way is also fine.

rschu1ze committed:
chore: incorporate review feedback

rschu1ze commented on programs/library-bridge/LibraryBridgeHandlers.cpp:
```diff
@@ -399,4 +410,202 @@ void ExternalDictionaryLibraryBridgeExistsHandler::handleRequest(HTTPServerReque
+            NativeWriter native_writer{string_write_buf, 0, block_write};
+            native_writer.write(block_write);
+
+            writeStringBinary(string_write_buf.str(), out);
```
I am actually not sure in what way `writeData` is more efficient than this code.

I suppose, since the exchanged data is typically not very big (it is ultimately limited by the maximum URL length) and since the data is cleared after each call to SQL function `catboostEvaluate`, we can do without further optimization for now.

kssenii replied:
> I am actually not sure in what way writeData is more efficient than this code.

currently you write the block twice - first you put it into string via WriteBufferFromOwnString and then write it to `out`. But you could write block to `out` at once without writing it into string.

rschu1ze replied:
Done, thanks.

rschu1ze committed:
perf: don't copy data around unnecessarily

kssenii reviewed (approved):
PS looks like you accidentally updated Nuraft submodules

rschu1ze commented:
Thanks for approving.

Yes, the nuraft update was an accident. I am currently squashing everything locally on top of a fresh master state, this will undo the update.

rschu1ze committed:
fix: test 02415_all_new_functions_must_be_documented

I am happy to move the documentation into the source code but currently
we'd need to maintain it twice (source code + traditional markdown docs)
because we don't render the docs from the source code.

rschu1ze committed:
fix: delete obsolete test 02413_model_evaluate_smoke
